### PR TITLE
Improve center nifti

### PR DESCRIPTION
**Pull request by arnaudmarcoux**
This PR aims to mitigate drawbacks of precendently merged center_nifti tool. 
We want to center the niftis of only specified modality in argument.

```
Optional arguments:
  --modality MODALITY, -m MODALITY
                        List of modalities you want to center the NIfTI images
                        (ex: "t1w fdg_pet dwi"). Default="t1w"
```

All the niftis that contains one of the keyword in their filename will be kept.

Wiki has been updated: https://github.com/aramis-lab/clinica/wiki/IO

A series of experiment have been conducted: we take a volume whose center is on the origin of the world coordinate system. We add an offset using coordinates of affine matrix [0, 3], [1, 3], [2, 3] (or by modifying the header['srow_x'][3], header['srow_y'][3], header['srow_z'][3], this is strictly equivalent).

Screen Shot 2019-10-14 at 18 24 37

It has been determined that volumes were still segmented with SPM when the L2 distance between origin and center of the volume did not exceed 100 mm. Above this distance, either the volume is either not segmented (SPM error), or the produced segmentation is wrong (not the shape of a brain anymore)

Screen Shot 2019-10-15 at 18 30 39
Screen Shot 2019-10-11 at 15 33 32

This discovery leads to 2 major modifications:
- At the beginning of `t1-volume, t1-volume-tissue-segmentation, t1-volume-existing-template`, we propose the following interface for clinica via the function `check_volume_location_in_world_coordinate_system`

Screen Shot 2019-10-16 at 14 56 25

- By default in the `center_nifti` tool, we only center files that are suceptible not to work with SPM, i.e center of the volume at maximum 80 mm of the origin of the world system coordinate.

Interface :  
Screen Shot 2019-10-18 at 10 59 41

### CODE REVIEW

File: clinica/iotools/utils/data_handling_cli.py [hunk 1]
```diff
@@ -117,31 +117,50 @@ def define_name(self):
...
 
     def define_options(self):
         self._args.add_argument("bids_directory",
-                                help='Path to the BIDS dataset directory in which you want to center the NIfTI files/images.')
+                                help='Path to the BIDS dataset directory in which you want to center the NIfTI '
+                                     + 'files/images.')
         self._args.add_argument("output_bids_directory",
-                                help='Path to the output directory. This is where your BIDS folder with centered nifti will appear.')
+                                help='Path to the output directory. This is where your BIDS folder with centered '
+                                     + 'nifti will appear.')
```
Comment (alexandreroutier) [inline clinica/iotools/utils/data_handling_cli.py]:
```suggestion
                                     + 'NIfTI files/images will appear.')
```

File: clinica/iotools/utils/data_handling_cli.py [hunk 2]
```diff
@@ -117,31 +117,50 @@ def define_name(self):
...
+        self._args.add_argument("--modality", '-m',
+                                help='List of modalities you want to center the NIfTI images (ex: "t1w fdg_pet dwi")'
+                                     + '. Default="t1w". All the files whose names contains one the keywords specified'
+                                     + ' in this list will be kept.',
+                                default='t1w')
+        self._args.add_argument('--center_all_files',
+                                help='Use this flag without argument to force to convert all NIFTI of the specified '
```
Comment (alexandreroutier) [inline clinica/iotools/utils/data_handling_cli.py]: Proposition:
```suggestion
                                help='Use this flag without argument to force the conversion of all NIfTI specified by the --modality flag. '
```

File: clinica/iotools/utils/data_handling_cli.py [hunk 3]
```diff
@@ -117,31 +117,50 @@ def define_name(self):
...
+                                     + 'modality(ies)',
+                                action='store_true',
+                                dest='center_all_files')
+        self._args.set_defaults(center_all_files=False)
```
Comment (alexandreroutier) [inline clinica/iotools/utils/data_handling_cli.py]: Shorter:

```python
        self._args.add_argument('--center_all_files', action='store_true', default=False,
                                help='Your description ending by a point.')
```
I use this e.g. in [DWI-Preprocessing CLI](https://github.com/alexandreroutier/clinica/blob/dwi_preprocessing/clinica/pipelines/dwi_preprocessing_using_t1/dwi_preprocessing_using_t1_cli.py)

File: clinica/iotools/utils/data_handling.py [hunk 4]
```diff
@@ -639,3 +655,298 @@ def center_all_nifti(bids_dir, output_dir):
         for error in all_errors:
             final_error_msg += '\n' + error
         raise RuntimeError(final_error_msg)
+    return nifti_files_filtered
+
+
+def write_list_of_files(file_list, output_file):
```
Comment (alexandreroutier) [inline clinica/iotools/utils/data_handling.py]: Needs at least a small description e.g. :
```suggestion
def write_list_of_files(file_list, output_file):
"""Save `file_list` list of files into `output_file` text file."""
```

File: clinica/iotools/utils/data_handling.py [hunk 5]
```diff
@@ -639,3 +655,298 @@ def center_all_nifti(bids_dir, output_dir):
...
+        return None
+
+    text_file = open(output_file, 'w+')
+    for created_file in file_list:
+        text_file.write(created_file + '\n')
+    text_file.close()
+    return output_file
+
+
+def check_volume_location_in_world_coordinate_system(nifti_list, bids_dir):
```
Comment (alexandreroutier) [inline clinica/iotools/utils/data_handling.py]: Needs small description

File: clinica/iotools/utils/data_handling.py [hunk 6]
```diff
@@ -639,3 +655,298 @@ def center_all_nifti(bids_dir, output_dir):
...
+        while True:
+            cprint('Your answer [yes/no]:')
+            answer = input()
+            if answer.lower() in ['yes', 'no']:
+                break
+            else:
+                cprint(Fore.RED + 'You must answer yes or no' + Fore.RESET)
+        if answer.lower() == 'no':
+            cprint(Fore.RED + 'Clinica will now exit...' + Fore.RESET)
+            sys.exit(0)
```
Comment (alexandreroutier) [inline clinica/iotools/utils/data_handling.py]: Is a timeout needed needed?

File: clinica/iotools/utils/data_handling.py [hunk 7]
```diff
@@ -639,3 +655,298 @@ def center_all_nifti(bids_dir, output_dir):
...
+    header['srow_x'][3], header['srow_y'][3], header['srow_z'][3], this is strictly equivalent).
+
+    It has been determined that volumes were still segmented with SPM when the L2 distance between origin and center of
+    the volume did not exceed 100 mm. Above this distance, either the volume is either not segmented (SPM error), or the
+    produced segmentation is wrong (not the shape of a brain anymore)
+
+    Args:
+        nii_volume: path to NIfTI volume
+        threshold_l2: maximum distance between origin of the world coordinate system and the center of the volume to
+                    be considered centered
```
Comment (alexandreroutier) [inline clinica/iotools/utils/data_handling.py]: Write here that 80 was determined empirically after several trials on <XX> dataset(s).

File: clinica/iotools/utils/data_handling_cli.py [hunk 8]
```diff
@@ -156,6 +175,29 @@ def run_command(self, args):
...
+        split_modality = [element for element in split_modality if element]
+
+        cprint('Clinica is now centering the requested images.')
+        centered_files = center_all_nifti(abspath(args.bids_directory),
+                                          abspath(args.output_bids_directory),
+                                          split_modality,
+                                          center_all_files=args.center_all_files)
+
+        # Write list of created files
+        timestamp = time.strftime("%Y%m%d-%H%M%S")
```
Comment (alexandreroutier) [inline clinica/iotools/utils/data_handling_cli.py]: If we want be homogenous with:
```python
        timestamp = strftime('%Y%m%d_%H%M%S', localtime(time()))
        tsv_file = '%s_subjects_sessions_list.tsv' % timestamp
```
in [get_subject_session_list](https://github.com/arnaudmarcoux/clinica/blob/improve_center_nifti/clinica/utils/io.py) function, you would need to change the formatting either in your function or [get_subject_session_list](https://github.com/arnaudmarcoux/clinica/blob/improve_center_nifti/clinica/utils/io.py).

File: clinica/iotools/utils/data_handling_cli.py [hunk 9]
```diff
@@ -117,31 +117,50 @@ def define_name(self):
...
+                                help='Path to the BIDS dataset directory in which you want to center the NIfTI '
+                                     + 'files/images.')
         self._args.add_argument("output_bids_directory",
-                                help='Path to the output directory. This is where your BIDS folder with centered nifti will appear.')
+                                help='Path to the output directory. This is where your BIDS folder with centered '
+                                     + 'nifti will appear.')
+        self._args.add_argument("--modality", '-m',
+                                help='List of modalities you want to center the NIfTI images (ex: "t1w fdg_pet dwi")'
+                                     + '. Default="t1w". All the files whose names contains one the keywords specified'
+                                     + ' in this list will be kept.',
```
Comment (alexandreroutier) [inline clinica/iotools/utils/data_handling_cli.py]: Proposition:
```python
    help=('List of modalities you want to center the NIfTI images (e.g.  --modality "t1w fdg_pet dwi"). '
          'All the files whose names contains one the keywords specified '
          ' in this list will be kept (default: --modality "t1w")'),
```
Only `(default: --modality "t1w").`should be written this way (cf CLI for each pipeline). The rest is only suggestion :)

**Comment by alexandreroutier** [comment]
Hi @arnaudmarcoux ,

I finished my feedback. I only had some typos comments. :)

Alexandre

File: clinica/iotools/utils/data_handling_cli.py (hunk 1 above)
Comment (arnaudmarcoux) [reply clinica/iotools/utils/data_handling_cli.py]: Done

File: clinica/iotools/utils/data_handling_cli.py (hunk 2 above)
Comment (arnaudmarcoux) [reply clinica/iotools/utils/data_handling_cli.py]: Done

File: clinica/iotools/utils/data_handling_cli.py (hunk 3 above)
Comment (arnaudmarcoux) [reply clinica/iotools/utils/data_handling_cli.py]: Done

File: clinica/iotools/utils/data_handling.py (hunk 4 above)
Comment (arnaudmarcoux) [reply clinica/iotools/utils/data_handling.py]: Done

File: clinica/iotools/utils/data_handling.py (hunk 5 above)
Comment (arnaudmarcoux) [reply clinica/iotools/utils/data_handling.py]: Done

File: clinica/iotools/utils/data_handling.py (hunk 6 above)
Comment (arnaudmarcoux) [reply clinica/iotools/utils/data_handling.py]: I think it is really important that the user see this message. If we launch the pipeline anyway, the user may never read this message and struggle to understand why his segmentations have failed.
I'd rather leave it that way

File: clinica/iotools/utils/data_handling.py (hunk 7 above)
Comment (arnaudmarcoux) [reply clinica/iotools/utils/data_handling.py]: Done

File: clinica/iotools/utils/data_handling.py [hunk 10]
```diff
@@ -639,3 +655,298 @@ def center_all_nifti(bids_dir, output_dir):
...
+                                                                                                      l2)
+
+        warning_message += '\nClinica provides a tool to counter this problem by replacing the center of the volume' \
+                           + ' at the origin of the world coordinates.\nUse the following command line to correct the '\
+                           + 'header of the faulty NIFTI volumes in a new folder:\n' + Fore.RESET \
+                           + Fore.BLUE + '\nclinica iotools center-nifti ' + abspath(bids_dir) + ' ' \
+                           + abspath(bids_dir) + '_centered\n\n' + Fore.YELLOW \
+                           + 'You will find more information on the command by typing ' + Fore.BLUE \
+                           + 'clinica iotools center-nifti' + Fore.YELLOW + ' in the console.\nDo you still want to '\
+                           + 'launch the pipeline now ?' + Fore.RESET
```
[the remark this answers is not in the document]
Comment (arnaudmarcoux) [reply clinica/iotools/utils/data_handling.py]: Done

File: clinica/iotools/utils/data_handling_cli.py (hunk 8 above)
Comment (arnaudmarcoux) [reply clinica/iotools/utils/data_handling_cli.py]: Done

File: clinica/iotools/utils/data_handling_cli.py (hunk 9 above)
Comment (arnaudmarcoux) [reply clinica/iotools/utils/data_handling_cli.py]: Done

**Comment by arnaudmarcoux** [comment]
All comments have been addressed

File: clinica/iotools/utils/data_handling.py (hunk 6 above)
Comment (alexandreroutier) [reply clinica/iotools/utils/data_handling.py]: Ok.

**Pull request merged**